bigtable-hbase-parent/bigtable-hbase-2.x/src/main/java/com/google/cloud/bigtable/hbase2_x/BigtableAsyncAdmin.java

```diff
@@ -63,51 +78,148 @@
+        return;
+      }
+      if (disabledTables.contains(tableName)) {
+        cf.completeExceptionally(new TableNotEnabledException(tableName));
+        return;
+      }
+      disabledTables.add(tableName);
+      LOG.warn("Table " + tableName + " was disabled in memory only.");
+      cf.complete(null);
+    });
+
+    return cf;
   }
 
   @Override
   public CompletableFuture<Void> deleteTable(TableName tableName) {
-    throw new UnsupportedOperationException("deleteTable"); // TODO
+    return CompletableFuture.supplyAsync(() -> {
+      Builder deleteBuilder = DeleteTableRequest.newBuilder();
+      deleteBuilder.setName(bigtableInstanceName.toTableNameStr(tableName.getNameAsString()));
+      return deleteBuilder;
+    }).thenCompose(
+        d -> FutureUtils.toCompletableFuture(bigtableTableAdminClient.deleteTableAsync(d.build())))
+        .thenAccept(r -> {
+          disabledTables.remove(tableName);
+        });
   }
 
   @Override
   public CompletableFuture<Boolean> tableExists(TableName tableName) {
-    throw new UnsupportedOperationException("tableExists"); // TODO
+    return listTableNames(Optional.of(Pattern.compile(tableName.getNameAsString())), false)
+        .thenApply(r -> r.stream().anyMatch(e -> e.equals(tableName)));
   }
 
   @Override
   public CompletableFuture<List<TableName>> listTableNames(Optional<Pattern> tableNamePattern,
       boolean includeSysTables) {
-    throw new UnsupportedOperationException("listTables"); // TODO
+    return requestTableList().thenApply(r -> {
+      Stream<TableName> result;
+      if (tableNamePattern.isPresent()) {
+        result = r.stream().map(e -> bigtableInstanceName.toTableId(e.getName()))
+            .filter(e -> tableNamePattern.get().matcher(e).matches())
+            .map(e -> TableName.valueOf(e));
+      } else {
+        result = r.stream().map(e -> bigtableInstanceName.toTableId(e.getName()))
+            .map(e -> TableName.valueOf(e));
+      }
+      return result.collect(Collectors.toList());
+    });
   }
 
   @Override
   public CompletableFuture<List<TableDescriptor>> listTables(Optional<Pattern> tableNamePattern,
       boolean includeSysTables) {
-    throw new UnsupportedOperationException("listTables"); // TODO
+    return requestTableList().thenApply(r -> {
+      List<TableDescriptor> result = new ArrayList<>();
+      Boolean hasNonEmptyPattern = tableNamePattern.isPresent();
+      for (Table table : r) {
+        String tableName = bigtableInstanceName.toTableId(table.getName());
+        if (hasNonEmptyPattern) {
+          if (tableNamePattern.get().matcher(tableName).matches()) {
+            result.add(tableAdapter2x.adapt(table));
+          }
+        } else {
+          result.add(tableAdapter2x.adapt(table));
+        }
+      }
+      return result;
+    });
+  }
+
+  private CompletableFuture<List<Table>> requestTableList() {
+    return CompletableFuture.supplyAsync(() -> {
```

Comment:
can't the builder be built outside of the CompletableFuture?

Reply:
It can execute outside in the calling thread itself, and any penalty is most likely negligible. The supplyAsync chained with thenCompose seems to provide a decent way to make most of the code async.